### PR TITLE
cranelift: arm32 codegen

### PR DESCRIPTION
**Pull request by jmkrauz**
This commit adds arm32 Thumb2 code generation for some IR insts.
I have used machinst backend and followed Aarch64 implementation as an example.

Floating-point instructions are not supported, because regalloc
does not seem to allow to represent overlapping register classes,
which are needed by VFP/Neon.

Another lacking feature is support for I64 and I128 types.

An abi implementation is not complete, however it makes it possible
to run simple clif tests.

It may be a first step towards  #1173.

This PR contains instruction emission test and clif tests for IR insts.

I don't know well who could review this.

### CODE REVIEW

File: cranelift/codegen/src/regalloc/solver.rs [hunk 1]
```diff
@@ -1149,24 +1149,14 @@ mod tests {
...
-        let shared_builder = settings::builder();
-        let shared_flags = settings::Flags::new(shared_builder);
-
-        isa::lookup(triple!("arm"))
-            .ok()
-            .map(|b| b.finish(shared_flags))
-    }
+    // Arm32 `TargetIsa` is now `TargetIsaAdapter`, which does not hold any info
+    // about registers, so we directly access `INFO` from registers-arm32.rs.
+    include!(concat!(env!("OUT_DIR"), "/registers-arm32.rs"));
```
Comment (bjorn3) [inline cranelift/codegen/src/regalloc/solver.rs:1159]: Maybe use x86 or riscv instead?

Comment (jmkrauz) [reply cranelift/codegen/src/regalloc/solver.rs:1159]: I believe some of the tests here or in regalloc/pressure.rs use the fact that S, D and Q registers overlap. I am not sure if we can achieve something similar with x86 or risc-v registers.

**Comment by cfallin** [comment]
Wow -- this is an excellent surprise to start my day with. First of all, thanks, @jmkrauz, for putting in all of this work!

I'll review this in the next few days; I expect it will take a bit of time. One high-level comment regarding the register-overlap issue: we've thought about (and need to solve) it for other architectures too, e.g. x86-32, so I expect we'll have a better answer eventually. For now, a stopgap solution (for correctness) might be to simply allocate pairs as the fundamental unit (i.e. `D` regs), and just burn an extra 32-bit register for each `f32`, and punt on 128-bit vector support for now. Not ideal but it will get us to basic completeness at least. @julian-seward1 might have more thoughts on this?

**Comment by bnjbvr** [comment]
Agreed that we could review and merge this behind a flag, or just letting know that support for arm32 is experimental; we've done this for other backends, and we can work incrementally on supporting more opcodes etc.

**Comment by cfallin** [comment]
Hi @jmkrauz, it looks like you've refactored to use the common `abi_impl` machinery (thanks!). I discussed this PR briefly today with a few others and we agree that it might be best to merge now, given that I haven't resolved the multiple-register-value question yet (which this would need for `i64` values). (Sorry about that -- a lot on my plate lately!)  If that's alright with you, of course; does this seem OK or do you have plans to refine the PR more first?

One thing I think we should do, however, is to put the new backend behind a feature flag until it implements Wasm MVP; otherwise someone might try to build wasmtime on arm32 and hit confusing panics. Perhaps just `#[cfg(feature = "arm32")]` on the main `mod arm32` statement in `cranelift/codegen/src/isa/mod.rs`?

**Comment by jmkrauz** [comment]
Merging now is absolutely fine with me.
I have added additional commit in which I put ARM backend behind `experimental_arm32` feature flag. In order to avoid failing arm32 filetests with `experimental_arm32` disabled, I have also added small check in `cranelift/filetests/src/runone.rs`.

**Comment by cfallin** [comment]
Excellent, thanks! Looking forward to seeing this become a full backend!

**Pull request merged**